### PR TITLE
Fix infomax optimizer bug

### LINKED ISSUE #5048: Bug: InfoMax3DModular pretraining fails to update model3d parameters (3D network remains untrained)
In `deepchem/models/torch_models/gnn3d.py`, the `InfoMax3DModular` model implements the 3D Infomax pretraining algorithm. This algorithm relies on contrastive learning by maximizing the mutual information between a 2D GNN (`model2d`) and a 3D GNN (`model3d`).

However, there is a critical architectural flaw that prevents the 3D network from ever being trained. 

Because `InfoMax3DModular` inherits from `ModularTorchModel` (and subsequently `TorchModel`), the PyTorch optimizer is implicitly constructed using `self.model.parameters()`. But in `gnn3d.py`, `build_model()` explicitly returns *only* `model2d`:

```python
    def build_model(self):
        # FIXME For pretraining task, both model2d and model3d but the super class
        # can't handle two models for contrastive learning, hence we pass only model2d
        return self.components['model2d']
```

Because `model3d` is not included in the returned model, **its parameters are never registered with the PyTorch optimizer**. During `fit()`, the loss is computed correctly using `model3d` in `loss_func`, but when `.backward()` is called and the optimizer steps, **only `model2d` is updated**. 

As a result, `model3d` remains completely frozen at its random initialization. The 2D network is effectively learning to match random noise, which entirely defeats the purpose of the 3D Infomax pretraining algorithm.

## To Reproduce

1. Initialize `InfoMax3DModular` with `task='pretraining'`.
2. Capture a snapshot of the weights of `model3d` before training.
3. Call `model.fit(dataset)` on a sample dataset.
4. Check the weights of `model3d` after training — they will be identical to the pre-training weights, proving that no gradients were applied to the 3D network during the contrastive learning phase.

## Expected behavior

During the `pretraining` task, the optimizer should receive the parameters for both `model2d` and `model3d` so that the contrastive learning updates both networks simultaneously. 

## Environment

* OS: Any
* Python version: Any
* DeepChem version: Master branch (latest)
* RDKit version (optional): N/A
* TensorFlow version (optional): N/A
* PyTorch version (optional): Any
* Any other relevant information: This is a static architecture bug in `gnn3d.py`.

## Additional context

There are a couple of ways to fix this without breaking the `TorchModel` API:
1. **Wrapper Module**: Modify `build_model()` to return a thin `nn.Module` wrapper during pretraining. This wrapper would register both `model2d` and `model3d` as sub-modules (so `self.model.parameters()` yields the parameters of both), but its `forward()` method would simply route to `model2d`.
2. **Custom Optimizer Definition**: Override the optimizer initialization logic in `ModularTorchModel` or `InfoMax3DModular` to explicitly chain and pass both parameters: `itertools.chain(self.components['model2d'].parameters(), self.components['model3d'].parameters())`.
```

### PR DESCRIPTION
**Pull request by Ritesh-1127**
## Description

Fixes #5048

**Summary and Motivation:**
During `InfoMax3DModular` pretraining, the 3D network (`model3d`) parameters were being entirely ignored by the PyTorch optimizer. Because `build_model()` returned only `model2d`, the parent `ModularTorchModel` never registered the 3D weights. This caused the 3D network to remain completely frozen at initialization, defeating the purpose of the contrastive learning algorithm.

To resolve this, I implemented a lightweight `_PretrainingWrapper` (`nn.Module`) that encapsulates both models. When instantiated for pretraining, `build_model()` now returns this wrapper. This ensures PyTorch's native `.parameters()` sweep captures the computational graphs for both networks so they update simultaneously. To maintain API compatibility, the wrapper routes the `forward()` pass directly through `model2d`.

I also added a unit test (`test_infomax3d_pretraining_parameter_registration`) in `test_gnn3d.py` that explicitly asserts both sets of network parameters are successfully registered to the main wrapper's computational graph.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Ritesh-1127** [comment]
Hi team, I have implemented a fix to ensure that the _PretrainingWrapper correctly exposes the 3D network parameters to the optimizer. The local tests are passing cleanly on my end. I’d really appreciate a review whenever you have a chance, and I'm happy to make any adjustments or changes if needed!

**Comment by Ritesh-1127** [comment]
Hi @shreyasvinaya, checking in on this! I implemented a _PretrainingWrapper to fix the PyTorch optimizer parameter registration issue for the 3D network. Could you approve the workflow runs when you have a moment so we can verify the tests? Thanks!

File: deepchem/models/torch_models/gnn3d.py [hunk 1]
```diff
@@ -565,16 +591,14 @@ def build_components(self):
 
     def build_model(self):
         """
-        Build the InfoMax3DModular model. This is the 2D network which is meant to be used for inference.
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/gnn3d.py:568]: Please revert this back to deepchem style docstrings

Comment (Ritesh-1127) [reply deepchem/models/torch_models/gnn3d.py:568]: Reverted!

**Comment by Ritesh-1127** [comment]
Hi @shreyasvinaya, all set! I reverted the docstring back to the DeepChem style and fixed the PyTorch marker and import ordering issues caught by the linter. Let me know if there's anything else needed!

**Review by Shriprasad-P — COMMENTED** [review]
## Review

**PR:** Fix infomax optimizer bug

**Touched:** `deepchem/models/tests/test_gnn3d.py`, `deepchem/models/torch_models/gnn3d.py`

- Addresses a concrete correctness/reliability issue based on the title and diff.
- Includes or touches tests — good for locking the behavior.
- Size looks manageable (+69/-5).

Commenting as a drive-by reviewer after reading the diff. Happy to look again if maintainers want a deeper pass on a specific file.
